File: Root Finding Problem/Methods/secantHybrid.py
# Jake Daniels
# MATH 4610

import numpy as np

def fn():
    f = input("Enter function you want to find the roots for (in terms of x): ")
    return f

def f(fn, x):
    fval = eval(fn)
    return fval

def tolerance():
    tol = eval(input("Input tolerance here: "))
    return tol

def maxIter():
    mI = eval(input("Input max number of iterations here: "))
    return mI

def wantTable():
    print("Do you want a table?")
    print("\t1) Yes")
    print("\t2) No")
    choice = eval(input("Enter here: "))
    return choice

def table(a,b,c):
    print("(1)" , a, " (2) ", b, " (3) ", c)

def secantHybrid():
    fun = fn()
    a, b = eval(input("Enter interval for function as (a,b) where b > a: "))
    tol = tolerance()
    mI = maxIter()
    # choice = wantTable()
    x0, x1 = a, b
    error = 10 * tol
    iter = 0
    f0 = f(fun, x0)
    f1 = f(fun, x1)
    while error > tol and iter < mI:
        x2 = x1 - ((f1 * (x1 - x0)) / (f1 - f0))
        secError = np.abs(x2 - x1)
        if secError > error:
            fa = f0
            fb = f1
            for i in range(1, 4):
                c = .5 * (a + b)
                fc = f(fun, c)
                if fa * fc < 0:
                    b = c
                    fb = fc
                elif fb * fc < 0:
                    a = c
                    fa = fc
            error = np.abs(b - a)
            x0, x1 = a, b
        else:
            x0 = x1
            x1 = x2
            f0 = f1
            f1 = f(fun, x2)
            error = secError
            iter = iter + 1
    print("Root approximation: ", x1)

File: RF Shared Library/secant_hybrid.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double secantHybrid();
double fval(double);

double secantHybrid(double (*f)(), double a, double b, double tol, double maxIter)
{
    double x0 = a;
    double x1 = b;
    double error = fabs(b-a);
    double iter = 0.0;
    double f0 = f(x0);
    double f1 = f(x1);
    while(error > tol && iter < maxIter)
    {
        double x2 = x1 - ((f1 * (x1 - x0)) / (f1 - f0));
        double secError = fabs(x2 - x1);
        double f2 = f(x2);
        double pos = f2 * f1;
        if(secError > error && pos < 0.0)
        {
            a = x1;
            b = x2;
            double fa = f0;
            double fb = f1;
            double c;
            double fc;
            int k = 4.0;
            for(int i=0; i<k; i++)
            {
                c = 0.5 * ( a + b );
                fc = f(c);
                if(fa*fc < 0.0)
                {
                    b = c;
                    fb = fc;
                } if(fb*fc < 0.0) {
                    a = c;
                    fa = fc;
                } else {
                    pos = 1.0;
                }
            }
            error = fabs(b - a);
            iter = iter + 1.0;
            x0 = a;
            x1 = b;
            f0 = fa;
            f1 = fb;
        } else {
            x0 = x1;
            x1 = x2;
            f0 = f1;
            f1 = f(x1);
            error = secError;
            iter = iter + 1.0;
        }
    }
    return x1;
}
